### PR TITLE
fix(gatsby): filter null values in headComponents, preBodyComponents and postBodyComponents

### LINKED ISSUE #12553: [gatsby] headComponents can be empty (null or empty array) values
## Description

This is related to #12549 and as discussed in the comments, sanitizing `headComponents` would ensure that plugins that do similar transformations as the one in that issue can assume that all elements are actual components.

### Steps to reproduce

As of now it is possible to call `setHeadComponents` or `replaceHeadComponents` with any values and there are no sanity checks in place.

### Expected result

Empty values (null, empty array) should be filtered out.

### Actual result

Everything gets added.

### Environment

Not environment specific.

### PR DESCRIPTION
**Pull request by danielkov**
## Description

> This is related to #12549 and as discussed in the comments, sanitizing headComponents would ensure that plugins that do similar transformations as the one in that issue can assume that all elements are actual components.
> -- <cite>#12553</cite>

## Related Issues

Fixes #12553

### CODE REVIEW

File: packages/gatsby/cache-dir/__tests__/static-entry.js [hunk 1]
```diff
@@ -61,6 +61,48 @@ const reverseHeadersPlugin = {
...
+    },
+  },
+}
+
+const checkNonEmptyHeadersPlugin = {
+  plugin: {
+    onPreRenderHTML: ({ getHeadComponents }) => {
+      const headComponents = getHeadComponents()
+      expect(headComponents.includes(null)).toBeFalsy()
+      expect(headComponents.find(val => Array.isArray(val) && val.length === 0))
```
Comment (pieh) [inline packages/gatsby/cache-dir/__tests__/static-entry.js]: this seems to miss assertion? the `expect()` should be falsy here as well, right?

File: packages/gatsby/cache-dir/__tests__/static-entry.js [hunk 2]
```diff
@@ -147,6 +189,42 @@ describe(`static-entry`, () => {
     })
   })
 
+  test(`onPreRenderHTML can filter out null value`, done => {
+    global.plugins = [setNullHeaderPlugin, checkNonEmptyHeadersPlugin]
+
+    StaticEntry(`/about/`, (_, html) => {
+      expect(html).toMatchSnapshot()
```
Comment (pieh) [inline packages/gatsby/cache-dir/__tests__/static-entry.js]: do we need snapshot testing here if we assert head components directly? I would like avoid busy snapshots like that that, especially that I think this snapshot would pass right now without any changes in `static-entry` file, right?

**Comment by pieh** [comment]
It would be great to apply that to `PreBodyComponents` and `PostBodyComponents` as well (would be nice to structure test to not repeat mocked plugins / tests for each of those

File: packages/gatsby/cache-dir/__tests__/static-entry.js (hunk 1 above)
Comment (danielkov) [reply packages/gatsby/cache-dir/__tests__/static-entry.js]: Good catch! It's a midnight-coding error on my part.

File: packages/gatsby/cache-dir/__tests__/static-entry.js (hunk 2 above)
Comment (danielkov) [reply packages/gatsby/cache-dir/__tests__/static-entry.js]: As it stands they do seem to be redundant. I will remove them.

**Comment by danielkov** [comment]
> It would be great to apply that to PreBodyComponents and PostBodyComponents as well

I will do that. I've scanned the official plugins and none of them seems to rely on the current behaviour. However I will browse through some of the popular plugins outside of this repo, to make sure it doesn't regress community packages.

File: packages/gatsby/cache-dir/static-entry.js [hunk 3]
```diff
@@ -49,6 +49,16 @@ const getPage = path => pagesObjectMap.get(path)
 
 const createElement = React.createElement
 
+const sanitizeHeadComponents = components => {
```
Comment (wardpeet) [inline packages/gatsby/cache-dir/static-entry.js]: you could rewrite this function to something like

```js
const sanitizeHeadComponents = components => {
  const componentsAsArray = Array.isArray(components) ? components : [components];

  return componentsAsArray.filter(Boolean);
};
```
The filter Boolean removes all falsy (undefined, null, fasle) values. I don't think you can have nested arrays.

Comment (danielkov) [reply packages/gatsby/cache-dir/static-entry.js]: As of right now in my particular example I can get the following, just using Helmet:

```js
[
  [
    {
      $$typeof: Symbol(react.element),
      type: "title",
      key: "",
      ref: null,
      props: [Object],
      _owner: null,
      _store: {}
    }
  ],
  [],
  [],
  [],
  [],
  [],
  null,
  {
    $$typeof: Symbol(react.element),
    type: "link",
    key: "gatsby-plugin-feed-0",
    ref: null,
    props: {
      rel: "alternate",
      type: "application/rss+xml",
      title: undefined,
      href: "/rss.xml"
    },
    _owner: null,
    _store: {}
  },
  // Other component definitions
];
```

They're all valid React Children, so it makes sense. I wanted to use this as a starting point for sanitizing the input.

Comment (wardpeet) [reply packages/gatsby/cache-dir/static-entry.js]: ok, I missed what you were trying to do. Thanks for clarifying.

File: packages/gatsby/cache-dir/static-entry.js [hunk 4]
```diff
@@ -49,6 +49,16 @@ const getPage = path => pagesObjectMap.get(path)
...
+  if (Array.isArray(components)) {
+    // remove falsy items
+    return components.filter(val => (Array.isArray(val) ? val.length > 0 : val))
+  } else {
```
Comment (wardpeet) [inline packages/gatsby/cache-dir/static-entry.js]: you can drop this else because of the return

**Pull request merged**